File: apps/web/src/api/analysis.ts
// API client for Phase 2 analysis endpoints

import type { PaginatedResponse } from "./client";

const BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "/api/v1";

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const url = `${BASE_URL}${path}`;
  const res = await fetch(url, {
    ...init,
    headers: { "Content-Type": "application/json", ...init?.headers },
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${body || res.statusText}`);
  }
  return res.json() as Promise<T>;
}

// ---- Types ----

export interface AnalysisSession {
  id: string;
  cycle_id: string;
  charter_id: string;
  paper_card_id: string;
  status: string;
  budget: Record<string, unknown> | null;
  stats: Record<string, unknown> | null;
  step_log: Array<Record<string, unknown>> | null;
  report_artifact_path: string | null;
  error: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface IngestedDocument {
  id: string;
  analysis_session_id: string;
  paper_card_id: string;
  fetch_method: string;
  source_url: string;
  normalized_sections: Array<Record<string, unknown>> | null;
  normalized_figures: Array<Record<string, unknown>> | null;
  normalized_tables: Array<Record<string, unknown>> | null;
  normalized_equations: Array<Record<string, unknown>> | null;
  quality_assessment: Record<string, unknown> | null;
  fetch_duration_ms: number | null;
  content_hash: string;
  created_at: string;
}

export interface PaperChunk {
  id: string;
  analysis_session_id: string;
  paper_card_id: string;
  chunk_type: string;
  section_path: string | null;
  ordinal: number;
  content: string;
  content_hash: string;
  chunk_metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface GraphNode {
  id: string;
  analysis_session_id: string;
  paper_card_id: string;
  node_type: string;
  label: string;
  description: string | null;
  properties: Record<string, unknown> | null;
  provenance: Record<string, unknown> | null;
  created_at: string;
}

export interface GraphEdge {
  id: string;
  analysis_session_id: string;
  source_node_id: string;
  target_node_id: string;
  edge_type: string;
  properties: Record<string, unknown> | null;
  provenance: Record<string, unknown> | null;
  confidence: number | null;
  created_at: string;
}

export interface CoverageDiagnostic {
  id: string;
  analysis_session_id: string;
  section_coverage: Record<string, unknown> | null;
  figure_coverage: Record<string, unknown> | null;
  table_coverage: Record<string, unknown> | null;
  equation_coverage: Record<string, unknown> | null;
  unlinked_artifacts: string[] | null;
  warnings: string[] | null;
  overall_score: number;
  created_at: string;
}

export interface PaperAnalysisPacket {
  id: string;
  analysis_session_id: string;
  paper_card_id: string;
  charter_id: string;
  summary: string;
  key_contributions: string[] | null;
  methods_used: Array<Record<string, unknown>> | null;
  datasets_referenced: Array<Record<string, unknown>> | null;
  reproducibility_notes: Record<string, unknown> | null;
  graph_summary: Record<string, unknown> | null;
  coverage_snapshot: Record<string, unknown> | null;
  chunk_count: number;
  node_count: number;
  edge_count: number;
  analysis_depth: string;
  created_at: string;
  updated_at: string;
}

export interface PaperReviewArtifact {
  id: string;
  analysis_packet_id: string;
  paper_card_id: string;
  strengths: string[] | null;
  weaknesses: string[] | null;
  open_questions: string[] | null;
  critique: string | null;
  scores: Record<string, number> | null;
  reading_priority: string | null;
  created_at: string;
}

export interface EvidenceCard {
  id: string;
  charter_id: string;
  cycle_id: string;
  paper_card_id: string;
  analysis_packet_id: string | null;
  evidence_type: string;
  claim: string;
  supporting_text: string | null;
  source_chunk_ids: string[] | null;
  source_graph_node_ids: string[] | null;
  confidence: number;
  analysis_depth: string;
  contradiction_flags: Record<string, unknown> | null;
  redundancy_group: string | null;
  extra_metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface AnalysisSessionStartResponse {
  session: AnalysisSession;
  job_id: string;
}

export interface QARequest {
  question: string;
  max_chunks?: number;
  expand_graph?: boolean;
}

export interface QAResponse {
  answer: string;
  supporting_chunks: Array<Record<string, unknown>>;
  supporting_nodes: Array<Record<string, unknown>>;
  confidence: number;
}

export interface LocateRequest {
  entity_type: string;
  query: string;
}

export interface LocateResponse {
  matches: Array<Record<string, unknown>>;
}

export interface AnalysisReport {
  session_id: string;
  markdown: string | null;
  json: Record<string, unknown> | null;
}

// ---- API functions ----

export function startAnalysis(
  paperCardId: string,
  charterId: string,
  cycleId: string,
  budget: Record<string, unknown> = {},
) {
  return apiFetch<AnalysisSessionStartResponse>(
    `/papers/${paperCardId}/analyze?charter_id=${charterId}&cycle_id=${cycleId}`,
    { method: "POST", body: JSON.stringify({ budget }) },
  );
}

export function fetchAnalysisSessions(opts: {
  charter_id?: string;
  cycle_id?: string;
  status?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.charter_id) params.set("charter_id", opts.charter_id);
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.status) params.set("status", opts.status);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<AnalysisSession>>(
    `/analysis${qs ? `?${qs}` : ""}`,
  );
}

export function fetchAnalysisSession(sessionId: string) {
  return apiFetch<AnalysisSession>(`/analysis/${sessionId}`);
}

export function fetchIngestedDocument(sessionId: string) {
  return apiFetch<IngestedDocument>(`/analysis/${sessionId}/document`);
}

export function fetchChunks(sessionId: string, opts: {
  chunk_type?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.chunk_type) params.set("chunk_type", opts.chunk_type);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<PaperChunk>>(
    `/analysis/${sessionId}/chunks${qs ? `?${qs}` : ""}`,
  );
}

export function fetchGraphNodes(sessionId: string, opts: {
  node_type?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.node_type) params.set("node_type", opts.node_type);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<GraphNode>>(
    `/analysis/${sessionId}/graph/nodes${qs ? `?${qs}` : ""}`,
  );
}

export function fetchGraphEdges(sessionId: string, opts: {
  edge_type?: string;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.edge_type) params.set("edge_type", opts.edge_type);
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<GraphEdge>>(
    `/analysis/${sessionId}/graph/edges${qs ? `?${qs}` : ""}`,
  );
}

export function fetchCoverage(sessionId: string) {
  return apiFetch<CoverageDiagnostic>(`/analysis/${sessionId}/coverage`);
}

export function fetchAnalysisPacket(paperCardId: string) {
  return apiFetch<PaperAnalysisPacket>(
    `/papers/${paperCardId}/analysis-packet`,
  );
}

export function fetchReviewArtifact(paperCardId: string) {
  return apiFetch<PaperReviewArtifact>(`/papers/${paperCardId}/review`);
}

export function askAnalysisQuestion(paperCardId: string, body: QARequest) {
  return apiFetch<QAResponse>(`/papers/${paperCardId}/qa`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function locateAnalysisEntity(paperCardId: string, body: LocateRequest) {
  return apiFetch<LocateResponse>(`/papers/${paperCardId}/locate`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function fetchAnalysisReport(sessionId: string) {
  return apiFetch<AnalysisReport>(`/analysis/${sessionId}/report`);
}

export function fetchEvidenceCards(opts: {
  charter_id?: string;
  cycle_id?: string;
  evidence_type?: string;
  min_confidence?: number;
  offset?: number;
  limit?: number;
} = {}) {
  const params = new URLSearchParams();
  if (opts.charter_id) params.set("charter_id", opts.charter_id);
  if (opts.cycle_id) params.set("cycle_id", opts.cycle_id);
  if (opts.evidence_type) params.set("evidence_type", opts.evidence_type);
  if (opts.min_confidence !== undefined)
    params.set("min_confidence", String(opts.min_confidence));
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  const qs = params.toString();
  return apiFetch<PaginatedResponse<EvidenceCard>>(
    `/evidence${qs ? `?${qs}` : ""}`,
  );
}

export function fetchEvidenceCard(evidenceId: string) {
  return apiFetch<EvidenceCard>(`/evidence/${evidenceId}`);
}

File: libs/schemas/analysis.py
"""Pydantic schemas for the Phase 2 analysis pipeline."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Ingestion DTOs
# ---------------------------------------------------------------------------


class NormalizedSection(BaseModel):
    """A section extracted from the paper's full text."""

    heading: str
    level: int = Field(ge=0, le=10)
    content: str
    page_range: list[int] | None = None


class NormalizedFigure(BaseModel):
    """A figure extracted from the paper."""

    id: str
    caption: str
    page: int | None = None
    content_description: str | None = None


class NormalizedTable(BaseModel):
    """A table extracted from the paper."""

    id: str
    caption: str
    page: int | None = None
    content_description: str | None = None


class NormalizedEquation(BaseModel):
    """An equation extracted from the paper."""

    id: str
    latex: str
    context: str | None = None
    page: int | None = None


class QualityAssessment(BaseModel):
    """Quality assessment for an ingested document."""

    structure_preserved: bool
    section_count: int = 0
    figure_count: int = 0
    table_count: int = 0
    equation_count: int = 0
    quality_score: float = Field(ge=0.0, le=1.0)
    quality_warnings: list[str] = Field(default_factory=list)


class IngestedDocumentRead(BaseModel):
    """Read schema for a persisted full-text artifact."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    paper_card_id: UUID
    fetch_method: str
    source_url: str
    normalized_sections: list[dict[str, Any]] | None
    normalized_figures: list[dict[str, Any]] | None
    normalized_tables: list[dict[str, Any]] | None
    normalized_equations: list[dict[str, Any]] | None
    quality_assessment: dict[str, Any] | None
    fetch_duration_ms: int | None
    content_hash: str
    created_at: datetime


# ---------------------------------------------------------------------------
# Analysis session
# ---------------------------------------------------------------------------


class AnalysisBudget(BaseModel):
    """Per-session analysis budget controls."""

    max_chunks: int = Field(default=500, ge=1, le=5000)
    graph_extraction_concurrency: int = Field(default=4, ge=1, le=16)
    evidence_extraction_concurrency: int = Field(default=4, ge=1, le=16)
    html_quality_threshold: float = Field(default=0.5, ge=0.0, le=1.0)


class AnalysisSessionStartRequest(BaseModel):
    """Body posted to ``POST /papers/{paper_card_id}/analyze``."""

    budget: AnalysisBudget = Field(default_factory=AnalysisBudget)


class AnalysisSessionRead(BaseModel):
    """Read schema for an analysis session."""

    model_config = {"from_attributes": True}

    id: UUID
    cycle_id: UUID
    charter_id: UUID
    paper_card_id: UUID
    status: str
    budget: dict[str, Any] | None
    stats: dict[str, Any] | None
    step_log: list[Any] | None
    report_artifact_path: str | None
    error: str | None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    completed_at: datetime | None


class AnalysisSessionStartResponse(BaseModel):
    """Returned by ``POST /papers/{paper_card_id}/analyze``."""

    session: AnalysisSessionRead
    job_id: UUID


# ---------------------------------------------------------------------------
# Chunks
# ---------------------------------------------------------------------------


class PaperChunkRead(BaseModel):
    """Read schema for a paper chunk."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    paper_card_id: UUID
    chunk_type: str
    section_path: str | None
    ordinal: int
    content: str
    content_hash: str
    chunk_metadata: dict[str, Any] | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Graph
# ---------------------------------------------------------------------------


class GraphNodeRead(BaseModel):
    """Read schema for a graph node."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    paper_card_id: UUID
    node_type: str
    label: str
    description: str | None
    properties: dict[str, Any] | None
    provenance: dict[str, Any] | None
    created_at: datetime


class GraphEdgeRead(BaseModel):
    """Read schema for a graph edge."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    source_node_id: UUID
    target_node_id: UUID
    edge_type: str
    properties: dict[str, Any] | None
    provenance: dict[str, Any] | None
    confidence: float | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Coverage
# ---------------------------------------------------------------------------


class CoverageDetail(BaseModel):
    """Coverage stats for one artifact type (sections, figures, etc.)."""

    total: int = 0
    covered: int = 0
    missing: list[str] = Field(default_factory=list)


class CoverageDiagnosticRead(BaseModel):
    """Read schema for coverage diagnostics."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    section_coverage: dict[str, Any] | None
    figure_coverage: dict[str, Any] | None
    table_coverage: dict[str, Any] | None
    equation_coverage: dict[str, Any] | None
    unlinked_artifacts: list[str] | None
    warnings: list[str] | None
    overall_score: float
    created_at: datetime


# ---------------------------------------------------------------------------
# Analysis packet
# ---------------------------------------------------------------------------


class PaperAnalysisPacketRead(BaseModel):
    """Read schema for the canonical analysis packet."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_session_id: UUID
    paper_card_id: UUID
    charter_id: UUID
    summary: str
    key_contributions: list[str] | None
    methods_used: list[dict[str, Any]] | None
    datasets_referenced: list[dict[str, Any]] | None
    reproducibility_notes: dict[str, Any] | None
    graph_summary: dict[str, Any] | None
    coverage_snapshot: dict[str, Any] | None
    chunk_count: int
    node_count: int
    edge_count: int
    analysis_depth: str
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Review artifact
# ---------------------------------------------------------------------------


class PaperReviewArtifactRead(BaseModel):
    """Read schema for the advisory review artifact."""

    model_config = {"from_attributes": True}

    id: UUID
    analysis_packet_id: UUID
    paper_card_id: UUID
    strengths: list[str] | None
    weaknesses: list[str] | None
    open_questions: list[str] | None
    critique: str | None
    scores: dict[str, Any] | None
    reading_priority: str | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Evidence cards
# ---------------------------------------------------------------------------


class ContradictionFlag(BaseModel):
    """One contradiction entry linking to another evidence card."""

    evidence_id: UUID
    reason: str
    model_confidence: float = Field(ge=0.0, le=1.0)


class EvidenceCardRead(BaseModel):
    """Read schema for an evidence card."""

    model_config = {"from_attributes": True}

    id: UUID
    charter_id: UUID
    cycle_id: UUID
    paper_card_id: UUID
    analysis_packet_id: UUID | None
    evidence_type: str
    claim: str
    supporting_text: str | None
    source_chunk_ids: list[str] | None
    source_graph_node_ids: list[str] | None
    confidence: float
    analysis_depth: str
    contradiction_flags: dict[str, Any] | None
    redundancy_group: str | None
    extra_metadata: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime


class EvidenceCardCreate(BaseModel):
    """Body for manual evidence creation."""

    paper_card_id: UUID
    evidence_type: str = Field(
        pattern="^(finding|method_claim|dataset_availability|limitation|comparison)$",
    )
    claim: str = Field(min_length=1, max_length=4000)
    supporting_text: str | None = None
    confidence: float = Field(default=0.5, ge=0.0, le=1.0)
    analysis_depth: str = Field(default="metadata", pattern="^(metadata|full)$")


# ---------------------------------------------------------------------------
# Graph-aware QA
# ---------------------------------------------------------------------------


class ChunkRef(BaseModel):
    """Reference to a supporting chunk in a QA answer."""

    chunk_id: UUID
    section_path: str | None = None
    ordinal: int | None = None
    snippet: str | None = None


class NodeRef(BaseModel):
    """Reference to a supporting graph node in a QA answer."""

    node_id: UUID
    node_type: str
    label: str


class QARequest(BaseModel):
    """Body for graph-aware QA."""

    question: str = Field(min_length=1, max_length=2000)
    max_chunks: int = Field(default=10, ge=1, le=50)
    expand_graph: bool = True


class QAResponse(BaseModel):
    """Response from graph-aware QA."""

    answer: str
    supporting_chunks: list[ChunkRef] = Field(default_factory=list)
    supporting_nodes: list[NodeRef] = Field(default_factory=list)
    confidence: float = Field(ge=0.0, le=1.0)


class LocateRequest(BaseModel):
    """Body for locating an entity in a paper."""

    entity_type: str = Field(pattern="^(concept|method|figure|table|equation|dataset)$")
    query: str = Field(min_length=1, max_length=500)


class LocateMatch(BaseModel):
    """One match from a locate query."""

    node: NodeRef
    chunks: list[ChunkRef] = Field(default_factory=list)


class LocateResponse(BaseModel):
    """Response from a locate query."""

    matches: list[LocateMatch] = Field(default_factory=list)


class AnalysisReportResponse(BaseModel):
    """Rendered analysis report bundle for a completed session."""

    model_config = ConfigDict(populate_by_name=True)

    session_id: UUID
    markdown: str | None = None
    json_payload: dict[str, Any] | None = Field(default=None, alias="json")
